src/coreclr/vm/multicorejitplayer.cpp

```diff
@@ -1349,53 +1106,149 @@ HRESULT MulticoreJitProfilePlayer::PlayProfile()
+            data2 = * (((const unsigned *) pBuffer) + 1);
+            rcdLen = data2 >> SIGNATURE_LENGTH_OFFSET;
+        }
+        else
+        {
+            hr = COR_E_BADIMAGEFORMAT;
+            break;
+        }
 
         if ((rcdLen > nSize) || (rcdLen & 3)) // Better DWORD align
         {
             hr = COR_E_BADIMAGEFORMAT;
+            break;
         }
-        else
+
+        if (rcdTyp == MULTICOREJIT_MODULE_RECORD_ID)
+        {
+            const ModuleRecord * pRec = (const ModuleRecord * ) pBuffer;
+
+            if (((unsigned)(pRec->lenModuleName
+                + pRec->lenAssemblyName
+                ) > (rcdLen - sizeof(ModuleRecord))) ||
+                (m_moduleCount >= m_headerModuleCount))
+            {
+                hr = COR_E_BADIMAGEFORMAT;
+            }
+            else
+            {
+                hr = HandleModuleRecord(pRec);
+            }
+        }
+        else if (rcdTyp == MULTICOREJIT_MODULEINF_RECORD_ID)
         {
-            if (rcdTyp == MULTICOREJIT_MODULE_RECORD_ID)
+            unsigned moduleIndex = data1 & (MAX_MODULES - 1);
+            unsigned level = (data1 & ((MAX_MODULE_LEVELS - 1) << MODULE_LEVEL_OFFSET)) >> MODULE_LEVEL_OFFSET;
+
+            hr = HandleModuleInfoRecord(moduleIndex, level);
+        }
+        else if (rcdTyp == MULTICOREJIT_METHODINF_RECORD_ID)
+        {
+            // Find all subsequent methods and jit/load them reversed if reversed order is required
+            bool reversedOrder = true;
+
+            if (reversedOrder)
             {
-                const ModuleRecord * pRec = (const ModuleRecord * ) pBuffer;
+                bool isMethod = true;
+                const BYTE * pCurBuf = pBuffer;
+                unsigned curSize = nSize;
+
+                unsigned sizes[MAX_WALKBACK] = {0};
+                int count = 0;
+
+                do
+                {
+                    unsigned curdata2 = * (((const unsigned *) pCurBuf) + 1);
+                    unsigned currcdLen = curdata2 >> SIGNATURE_LENGTH_OFFSET;
+
+                    if (currcdLen > curSize)
+                    {
+                        hr = COR_E_BADIMAGEFORMAT;
+                        break;
+                    }
+
+                    sizes[count] = currcdLen;
+                    count++;
+
+                    pCurBuf += currcdLen;
+                    curSize -= currcdLen;
+
+                    if (curSize == 0)
+                    {
+                        break;
+                    }
 
-                if (((unsigned)(pRec->lenModuleName
-                    + pRec->lenAssemblyName
-                    ) > (rcdLen - sizeof(ModuleRecord))) ||
-                    (m_moduleCount >= m_headerModuleCount))
+                    unsigned curdata1 = * (const unsigned *) pCurBuf;
+                    isMethod = RecorderInfo::isMethod(curdata1);
+                }
+                while (isMethod && count < MAX_WALKBACK);
+
+                _ASSERTE(count > 0);
```

Comment:
Looks like this could fail if `count` happened to be `0` and `hr == COR_E_BADIMAGEFORMAT` from above. Probably need  a `SUCCEEDED(hr)` here and in the `if` expression below.